From 074998b4def38bde393766729812ed445719c2b1 Mon Sep 17 00:00:00 2001
From: Yan Markman <ymarkman@marvell.com>
Date: Tue, 7 Aug 2018 17:31:26 +0300
Subject: [PATCH 0540/1051] net: mvpp2: gracefull stop-dev sequence

Under active traffic the BM/RX and TX PP2-HW could be non-empty.
This leads to "cleaning queue timed out", "Tx stop timed out"
and after these to Panic on interface-up.

FIX:
Stop asap new packets arriving from both RX and TX directions,
  keep egress enabled permitting to HW to send out pending packets,
  keep sw-interrupts enabled permitting to SW to handle and clear
  from last RX and TX-done packets.
Yield the context for graceful cleaning & finishing.
Continue with original Stop-sequence.

Change-Id: I20e664c1cc9df210600c6d36bdf85815cb1d2d54
Signed-off-by: Yan Markman <ymarkman@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/59470
Reviewed-by: Stefan Chulski <stefanc@marvell.com>
Tested-by: Stefan Chulski <stefanc@marvell.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c b/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
index 9f2dd1fdb82d..ff1250cae3a4 100644
--- a/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
+++ b/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
@@ -3317,7 +3317,16 @@ static void mvpp2_stop_dev(struct mvpp2_port *port)
 {
 	int i;
 
+	/* Stop-dev called by ifconfig but also by ethtool-features.
+	 * Under active traffic the BM/RX and TX PP2-HW could be non-empty.
+	 * Stop asap new packets ariving from both RX and TX directions,
+	 * but do NOT disable egress free/send-out and interrupts tx-done,
+	 * yeild and msleep this context for gracefull finishing.
+	 */
 	mvpp2_tx_stop_all_queues(port->dev);
+	mvpp2_ingress_disable(port);
+	msleep(40);
+	mvpp2_egress_disable(port);
 
 	/* Disable interrupts on all threads */
 	mvpp2_interrupts_disable(port);
-- 
2.17.1

